flax/training/checkpoints.py

```diff
@@ -89,92 +89,109 @@ def safe_normpath(path: str) -> str:
+      ckpt_dir = safe_normpath(ckpt_dir)
+      ckpt_tmp_path = _checkpoint_path(ckpt_dir, 'tmp', prefix)
+      ckpt_path = _checkpoint_path(ckpt_dir, step, prefix)
+      gfile.makedirs(os.path.dirname(ckpt_path))
+      base_path = os.path.join(ckpt_dir, prefix)
+      checkpoint_files = gfile.glob(base_path + '*')
+
+      if ckpt_path in checkpoint_files:
+        if not overwrite:
+          raise errors.InvalidCheckpointError(ckpt_path, step)
+      else:
+        checkpoint_files.append(ckpt_path)
+
+      checkpoint_files = natural_sort(checkpoint_files)
+      # Handle the case if the job was preempted after the temporary checkpoint was
+      # written, but before it was renamed to the final checkpoint name
+      if checkpoint_files[-1] == ckpt_tmp_path:
+        checkpoint_files.pop(-1)
+      if ckpt_path != checkpoint_files[-1]:
+        if not overwrite:
+          raise errors.InvalidCheckpointError(ckpt_path, step)
+
+      with gfile.GFile(ckpt_tmp_path, 'wb') as fp:
+        fp.write(serialization.to_bytes(target))
+
+      # Rename once serialization and writing finished.
+      gfile.rename(ckpt_tmp_path, ckpt_path, overwrite=overwrite)
+      logging.info('Saved checkpoint at %s', ckpt_path)
+
+      # Remove newer checkpoints
+      if overwrite:
+        ind = checkpoint_files.index(ckpt_path) + 1
+        newer_ckpts = checkpoint_files[ind:]
+        checkpoint_files = checkpoint_files[:ind]
+        for path in newer_ckpts:
+          logging.info('Removing checkpoint at %s', path)
+          try:
+            gfile.remove(path)
+          except errors.NotFoundError:
+            # Note this is likely to happen if this write was slower than the next one, but it doesn't matter if thread i or i+n deletes it.
+            logging.info('Removing checkpoint at %s failed, likely due to it being already deleted', path)
+
+      # Remove old checkpoint files.
+      if len(checkpoint_files) > keep:
+        old_ckpts = checkpoint_files[:-keep]
+        for path in old_ckpts:
+          logging.info('Removing checkpoint at %s', path)
+          try:
+            gfile.remove(path)
+          except errors.NotFoundError:
+            # Note this is likely to happen if this write was slower than the next one, but it doesn't matter if thread i or i+n deletes it.
+            logging.info('Removing checkpoint at %s failed, likely due to it being already deleted', path)
+
+      return ckpt_path
+
+    # Now we check if we are writing in blocking mode or not.
+    if blocking:
+      # if blocking is True, we will wait for the file write to be completed before returning.
+      return _save_checkpoint(**_kwargs)
+    else:
+      # if blocking is False, we will return the handle to the future.
+      # NOTE: we set the max_workers to 1 because the work is serial but we want to have it done asynchronously.
+      executor=thread.ThreadPoolExecutor(max_workers=1)
```

Comment:
Because you create the executor inside save_checkpoint calling it while the previous checkpoint is still being stored results in a race. I'm not convinced we should make this a keyword argument. Why not add this snippet with the executor in the save_checkpoint docstring with a note explaining users about the performance implications of blocking vs non-blocking storage? This way users can also use the thread pool for other blocking stuff like serializing a tf.data iterator, storing metrics, etc.

Reply:
@jheek 
On the doc string makes sense, will do.

On adding it as a keyword arg and the use of the executor, I'm following the design pattern of `restore_checkpoint` https://github.com/google/flax/blob/f458637584a460c139339e29fc0fe70c9ca15d49/flax/training/checkpoints.py#L253-L278 which adds the `parallel` keyword arg. I believe it needs to be a keyword argument because of the performance implications (mentioned in detail in #1543)

On the race, this is the result of the storage itself being slow to the point the compute of thread i+n might finish writing before thread i complete the write!... when the race happens the effect is thread i tries to delete an old checkpoint that doesn't exist anymore because thread i+n already did (which is not a bad thing , it doesn't matter who deletes checkpoint i-keep). Also this is handled in try-expect blocks around the gfile.remove https://github.com/google/flax/blob/1f7db1e19299655dd88c601c821c0bd129e62130/flax/training/checkpoints.py#L166-L181

But I see your point if this is used with the `overwrite=True` param.  I'll add more tests on my end and figure out how to support both `overwrite = True` and `overwrite=False`



thoughts?
Thanks again for the feedback!

Reply:
The benefit of parallel is that the multithreading is contained within the restore (restore_checkpoint itself is always blocking). Also for save_checkpoint we need to be a bit more careful because it is usually in the train loop.

It might be worth adding a blocking=False option if we can avoid the race condition indeed. Although I'm still a bit unsure if the additional complexity is worth it. Keeping the thread pool in the user's train loop has the added benefit of letting users add their own blocking code to the thread pool. It is also more explicit so it should make it easier to reason about the code and understand its performance.

Reply:
I see your point about the complexity vs. the user's training loop ... I suggest we add a "block until" in the imagenet example (because IIUC that is the "example") ala https://github.com/google/flax/blob/f458637584a460c139339e29fc0fe70c9ca15d49/examples/imagenet/train.py#L380-L381
so something like
```python
for step in steps:
    future = save_checkpoint(...)
    futures.append(future)
...
# Wait until all file writes are done before exiting
[f.result() for f in futures]
# Wait until computations are done before exiting 
jax.random.normal(jax.random.PRNGKey(0), ()).block_until_ready() 
```
even 10% decrease of your wall time on a v3-128 for the imagenet example is worth it IMHO 😅 
And mind you, the default behavior is `blocking=True` to make sure it is indeed backward compatible without any surprises.